lightning/src/ln/onion_utils.rs

```diff
@@ -1828,15 +1828,54 @@ where
 							trampoline_shared_secret,
 						),
 					}),
-					Ok((_, None)) => Err(OnionDecodeErr::Malformed {
+					Ok((msgs::InboundTrampolinePayload::BlindedForward(hop_data), None)) => {
+						if hop_data.intro_node_blinding_point.is_some() {
+							return Err(OnionDecodeErr::Relay {
+								err_msg: "Non-final intro node Trampoline onion data provided to us as last hop",
+								err_code: 0x4000 | 22,
+								shared_secret,
+								trampoline_shared_secret: Some(SharedSecret::from_bytes(
+									trampoline_shared_secret,
+								)),
+							});
+						}
+						Err(OnionDecodeErr::Malformed {
+							err_msg: "Non-final Trampoline onion data provided to us as last hop",
+							err_code: INVALID_ONION_BLINDING,
+						})
+					},
+					Ok((msgs::InboundTrampolinePayload::BlindedReceive(hop_data), Some(_))) => {
+						if hop_data.intro_node_blinding_point.is_some() {
+							return Err(OnionDecodeErr::Relay {
+								err_msg: "Final Trampoline intro node onion data provided to us as intermediate hop",
+								err_code: 0x4000 | 22,
+								shared_secret,
+								trampoline_shared_secret: Some(SharedSecret::from_bytes(
+									trampoline_shared_secret,
+								)),
+							});
+						}
+						Err(OnionDecodeErr::Malformed {
+							err_msg:
+								"Final Trampoline onion data provided to us as intermediate hop",
+							err_code: INVALID_ONION_BLINDING,
+						})
+					},
+					Ok((_, None)) => Err(OnionDecodeErr::Relay {
```

Comment:
Probably more clear to be explicit with these final two arms?
`Ok((msgs::InboundTrampolinePayload::Forward(_) | msgs::InboundTrampolinePayload::Receive(_), None)) => Err(OnionDecodeErr::Relay {
`

Reply:
Yes. I was wondering what cases I was looking at, what was left. I think explicit makes it easier?